src/lib/io/sol-ipm.c

```diff
@@ -0,0 +1,323 @@
+            receiver->last_remote_blob = remote_blob;
+            int r = sol_mainloop_event_post(&me);
+            return;
+        }
+    }
+
+    /* Is it really OK/safe to print from isr? */
+    SOL_INF("Got message with id %" PRIu32 " but there's no receiver associated to it",
+        id);
+}
+
+SOL_API int
+sol_ipm_set_receiver(uint32_t id,
+    void (*receive_cb)(void *cb_data, uint32_t id, struct sol_blob *message),
+    const void *data)
+{
+    int idx;
+    struct sol_ipm_receiver_entry *receiver;
+
+    SOL_INT_CHECK(id, > ipm_max_id, -EINVAL);
+    SOL_INT_CHECK(id, == 0, -EINVAL);
+
+    SOL_VECTOR_FOREACH_IDX (&ipm_receivers, receiver, idx) {
+        if (receiver->id == id) {
+            if (receive_cb) {
+                receiver->receive_cb = receive_cb;
+                return 0;
+            } else {
+                return sol_vector_del(&ipm_receivers, idx);
+            }
+        }
+    }
+
+    /* Not changing or deleting. Must have a receive_cb*/
+    SOL_NULL_CHECK(receive_cb, -EINVAL);
+
+    receiver = sol_vector_append(&ipm_receivers);
+    SOL_NULL_CHECK_MSG(receiver, -ENOMEM, "Could not add receiver to id %" PRIu32,
+        id);
+
+    receiver->id = id;
+    receiver->receive_cb = receive_cb;
+    receiver->data = data;
+
+    return 0;
+}
+
+SOL_API int
+sol_ipm_send(uint32_t id, struct sol_blob *message,
+    void (*message_consumed_cb)(void *cb_data, uint32_t id, struct sol_blob *message),
+    const void *data)
+{
+    int r;
+    struct sol_ipm_sent_message *sent;
+
+    SOL_INT_CHECK(id, > ipm_max_id, -EINVAL);
+    SOL_INT_CHECK(id, == 0, -EINVAL);
+    SOL_NULL_CHECK(message, -EINVAL);
+
+    sent = sol_vector_append(&ipm_sent_messages);
+    SOL_NULL_CHECK(sent, -ENOMEM);
+
+    r = ipm_send(ipm_sender, 1, id, &message, sizeof(message));
+    SOL_INT_CHECK_GOTO(r, > 0, fail);
+
+    sent->id = id;
+    sent->message_consumed_cb = message_consumed_cb;
+    sent->data = data;
+    sent->message = sol_blob_ref(message); /* Hold a reference to it so it doesn't die early*/
+
+    return 0;
+
+fail:
+    sol_vector_del_last(&ipm_sent_messages);
+    return -r;
+}
+
+
+SOL_API uint32_t
+sol_ipm_max_id(void)
+{
+    return ipm_max_id;
+}
+
+int
+sol_ipm_init(void)
+{
+    int max_data_size = 0, r;
+
+    sol_log_domain_init_level(SOL_LOG_DOMAIN);
+
+    ipm_receiver = device_get_binding("message_ipm_receiver");
+    SOL_NULL_CHECK_MSG(ipm_receiver, -EINVAL, "Could not get IPM receiver channel");
+
+    ipm_sender = device_get_binding("message_ipm_sender");
+    SOL_NULL_CHECK_MSG(ipm_sender, -EINVAL, "Could not get IPM sender channel");
+
+    max_data_size = ipm_max_data_size_get(ipm_sender);
+    if (max_data_size < sizeof(void *)) {
+        SOL_WRN("IPM max data size < sizeof(void *)");
+        return -EINVAL;
+    }
+
+    ipm_max_id = ipm_max_id_val_get(ipm_sender);
+
+    ipm_register_callback(ipm_receiver, ipm_receiver_cb, NULL);
+    r = ipm_set_enabled(ipm_receiver, 1);
+    if (r != 0) {
```

Comment:
I think you should only enable IPM if it's used, make the "lazy_init()" and shutdown pattern.

What do you think? Maybe it's not the case since one can disable the IPM, but I'd say enable by default and make its "non usage" cheap enough